OSASINFRA-3238: Improve API and Ingress VIPs validation

dkokkino opened the pull request:
• This commit enhances VIPs validation to ensure the primary IP family of the VIPs matches the primary IP family of all the network fields. 
• Code repurposed from prior closed PR (#7504)

dkokkino commented:
/retest

MaysaMacedo commented:
/cc @mkowalski @mandre

MaysaMacedo commented:
/test e2e-openstack-dualstack

mkowalski commented:
/test e2e-metal-ovn-dualstack

mkowalski commented:
/test e2e-metal-ipi-ovn-dualstack

mkowalski reviewed (changes requested):
I have a feeling the if-else logic here becomes a real spaghetti...

What we have now is

```
[1] if len(vips.API) == 0
[2] else if len(vips.API) <= 2
[3]    if len(vips.API) >= 1
[4]    if len(vips.API) == 2
```

and it seems to me like this could be simplified. For example [3] is redundant if you look and realize that you did not enter [1].

dkokkino commented:
> I have a feeling the if-else logic here becomes a real spaghetti...
> 
> What we have now is
> 
> ```
> [1] if len(vips.API) == 0
> [2] else if len(vips.API) <= 2
> [3]    if len(vips.API) >= 1
> [4]    if len(vips.API) == 2
> ```
> 
> and it seems to me like this could be simplified. For example [3] is redundant if you look and realize that you did not enter [1].

Thanks for pointing it out! I have removed if statement [3] and made sure the logic within that statement always runs if it enters [2]

mkowalski commented on pkg/types/validation/installconfig_test.go:
```diff
@@ -2109,27 +2135,27 @@ func TestValidateInstallConfig(t *testing.T) {
 			name: "baremetal API VIP set to an incorrect IP Family",
 			installConfig: func() *types.InstallConfig {
 				c := validInstallConfig()
-				c.Networking = validDualStackNetworkingConfig()
+				c.Networking = InvalidPrimaryV6DualStackNetworkingConfig()
 				c.Platform = types.Platform{
 					BareMetal: validBareMetalPlatform(),
 				}
 				c.Platform.BareMetal.APIVIPs = []string{"ffd0::"}
 				return c
 			}(),
-			expectedError: `platform.baremetal.apiVIPs: Invalid value: "ffd0::": VIP for the API must be of the same IP family with machine network's primary IP Family for dual-stack IPv4/IPv6`,
+			expectedError: `[platform.baremetal.apiVIPs: Invalid value: "ffd0::": serviceNetwork primary IP Family and primary IP family for the API VIP should match]`,
```
Please add your own test case instead of modifying the current one. They are not testing exactly the same combination, even though in today's logic of `validateAPIAndIngressVIPs` they run through the same code (tomorrow someone may modify the validation and one thing suddenly regresses and starts to pass)

Basically, your test is

- machine network primary v6
- service network primary v4
- cluster network primary v6
- api vip v6

and the old test is

- machine network primary v4
- service network primary v4
- cluster network primary v4
- api vip v6

dkokkino replied:
@mkowalski I added my own new tests instead of modifying the current ones as you suggested please let me know if you have any further advice.

mkowalski replied:
It is great that you added a new one! You only need now to fix the error message for existing test(s) because of your modification, i.e. look at the failure

```
Error:      	Expect "[platform.baremetal.ingressVIPs: Invalid value: "ffd0::": machineNetwork primary IP Family and primary IP family for the Ingress VIP should match, platform.baremetal.ingressVIPs: Invalid value: "ffd0::": serviceNetwork primary IP Family and primary IP family for the Ingress VIP should match, platform.baremetal.ingressVIPs: Invalid value: "ffd0::": clusterNetwork primary IP Family and primary IP family for the Ingress VIP should match]" to match "platform.baremetal.ingressVIPs: Invalid value: "ffd0::": VIP for the Ingress must be of the same IP family with machine network's primary IP Family for dual-stack IPv4/IPv6"
```

In the past there was only one error message for the mismatch. Now you append an error for every of machine, service, and cluster networks. Old test(s) do not know about this change.

You need to go one-by-one over whatever unit test(s) failed and fix expected error messages.

mkowalski commented:
/test e2e-metal-ipi-ovn-dualstack

mkowalski commented:
/test e2e-openstack-dualstack

mkowalski commented:
/test e2e-openstack-dualstack
/test e2e-metal-ipi-ovn-dualstack

MaysaMacedo commented:
@mkowalski hey, can you take another look at this PR?

MaysaMacedo commented:
@dkokkino looks like unit test is needs to be fixed. Can you take a look?

https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9438/pull-ci-openshift-installer-main-unit/1891530132485574656

mkowalski commented:
Functional change is good for me. I can't LGTM right now because fixing the tests will remove my label. I will add LGTM as soon as tests are fixed

dkokkino commented:
> @dkokkino looks like unit test is needs to be fixed. Can you take a look?
> 
> https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9438/pull-ci-openshift-installer-main-unit/1891530132485574656

Okay I believe i fixed the 2 failing unit tests.

mkowalski commented:
>Okay I believe i fixed the 2 failing unit tests.

There is still failure. If you scroll slightly down you will see red clickable `ci/prow/unit` that will give you the latest failure message

dkokkino commented:
> > Okay I believe i fixed the 2 failing unit tests.
> 
> There is still failure. If you scroll slightly down you will see red clickable `ci/prow/unit` that will give you the latest failure message

Is there a way for me to test this locally without having to trigger CI every time I want to test my changes? I have changed the 2 tests that were failing in ```ci/prow/unit```.

dkokkino commented:
/retest

MaysaMacedo commented:
/retest-required
/lgtm

dkokkino commented:
@mkowalski The unit tests are now passing

dkokkino commented:
/retest-required

mkowalski commented:
/approve

dkokkino commented:
/retest

dkokkino commented:
@pawanpinjarkar Could you please approve this PR when you have a chance? Thank you!

pawanpinjarkar commented:
@sadasu @patrickdillon can you please look into this as touches the core installer part.

sadasu commented:
/approve
